src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -8333,6 +8379,312 @@ private StatementSyntax ParseLocalDeclarationStatement()
+               || (this.CurrentToken.IsVar && this.PeekToken(1).Kind == SyntaxKind.OpenParenToken))
+            {
+                var resetPoint = this.GetResetPoint();
+
+                try
+                {
+                    var deconstruction = ParseDeconstructionDeclaration(withEquals);
+                    if ((object)deconstruction == null || !DeconstructionVariableLooksGood(deconstruction))
+                    {
+                        this.Reset(ref resetPoint);
+                        return null;
+                    }
+                    else
+                    {
+                        return deconstruction;
+                    }
+
+                }
+                finally
+                {
+                    this.Release(ref resetPoint);
+                }
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Parses a deconstruction-declaration, which can appear in a local-declaration statement or a for statement.
+        /// Returns null if this does not look like a deconstruction-declaration after all (equal sign missing).
+        ///
+        /// The syntax is `deconstruction-variables = expression` and can appear in local declarations and for statements.
+        /// </summary>
+        private VariableDeclarationSyntax ParseDeconstructionDeclaration(bool withEquals, bool topLevel = true)
+        {
+            if (this.CurrentToken.IsVar && this.PeekToken(1).Kind == SyntaxKind.OpenParenToken)
+            {
+                // parses `var (...) = expression`
+                return ParseDeconstructionVarForm(withEquals);
+            }
+            else if (this.CurrentToken.Kind == SyntaxKind.OpenParenToken)
+            {
+                // parses `(...) = expression` where the elements are deconstruction variables
+                return ParseDeconstructionListForm(withEquals);
+            }
+            else if (!topLevel)
+            {
+                return ParseDeconstructionTypeIdPart();
+            }
+            else
+            {
+                return null;
+            }
+        }
+
+        /// <summary>
+        /// Parses `(..., ..., ...)` where each part is deconstruction variables (such as `type id` or `var (..., ...)` or another list `(..., ..., ...)`).
+        /// </summary>
+        private VariableDeclarationSyntax ParseDeconstructionListForm(bool withEquals)
+        {
+            var list = _pool.AllocateSeparated<VariableDeclarationSyntax>();
+
+            try
+            {
+                var openParen = this.EatToken(SyntaxKind.OpenParenToken);
+
+                if (this.CurrentToken.Kind != SyntaxKind.CloseParenToken)
+                {
+                    var variable = ParseDeconstructionDeclaration(withEquals: false, topLevel: false);
+
+                    if ((object)variable != null)
+                    {
+                        list.Add(variable);
+                    }
```

Comment:
Can the loop below be rewritten to include this `ParseDeconstructionDeclaration(...); ... list.Add(variable);` and avoid the duplication?